src/metpy/xarray.py

```diff
@@ -1177,13 +1178,11 @@ def wrapper(*args, **kwargs):
 
             # Cast all Variables to their data and warn
             # (need to do before match finding, since we don't want to rewrap as Variable)
-            for arg_name in bound_args.arguments:
-                if isinstance(bound_args.arguments[arg_name], xr.Variable):
-                    warnings.warn(
-                        f'Argument {arg_name} given as xarray Variable...casting to its data. '
-                        'xarray DataArrays are recommended instead.'
-                    )
-                    bound_args.arguments[arg_name] = bound_args.arguments[arg_name].data
+            if _mutate_arguments(bound_args, xr.Variable, lambda arg: arg.data):
+                warnings.warn(
+                    'Argument given as xarray Variable...casting to its data. '
+                    'xarray DataArrays are recommended instead.'
+                )
```

Comment:
I feel like having the argument name that was provided as a `Variable` is rather useful for troubleshooting. Is there any feasible way to get that functionality back?

Reply:
Well, the goal with the refactor was to avoid having every place where we loop over arguments with names grow an additional loop over arguments without names. By using the same block/callable of code for all arguments, you lose the assumption of a name. I agree it would be nice, but I can't think of a way to get it back without increasing the code complexity here...

Reply:
Ok, I *think* I have something...